src/lib/snark_params/scan_state_constants.ml

```diff
@@ -0,0 +1,33 @@
+[%%import
+"../../config.mlh"]
+
+(* in the config files, we either have:
+
+    - transaction_capacity_log_2 and work_delay exp explicitly, or
+    - those same parameters, and a tps_goal
+
+   use the config boolean scan_state_with_tps_goal to distinguish those cases
+
+   If the tps_goal is given, then we can calculate a desired number of snark
+   workers.
+
+*)
+
+module type S = Scan_state_constants_intf.S
+
+[%%inject
+"transaction_capacity_log_2", scan_state_transaction_capacity_log_2]
+
+[%%inject
+"work_delay", scan_state_work_delay]
+
+[%%if
+scan_state_with_tps_goal]
+
+[%%inject
+"tps_goal", scan_state_tps_goal]
+
+let number_of_snark_workers () =
+  Int.pow 2 transaction_capacity_log_2 / work_delay
```

Comment:
surely this needs to use the `tps_goal` parameter somehow?

I still don't follow why the number of necessary snark workers decreases as work delay increases. At minimum, we need enough workers to produce (the number of `Transaction.t`s created per second) snarks per second, regardless of the data structure. There can't be a free lunch here. Plus whatever the merge overhead is. IIUC it's 2x, but I don't know the details.

Reply:
`transaction_capacity_log_2` is calculated using `tps_goal`? Regarding the inverse relation, as the delay increases there's more time available and therefore can do with less snark workers. Increasing the delay surely increases the number of available work but we don't have to do all of that to be able to achieve max throughput. We need snark workers to complete just enough work required to add `2^transaction_capacity_log_2` transactions which is `2^transaction_capacity_log_2` work bundles .

Reply:
I'm confused ... which parameters are fixed, and which are dependent?

If we provide `tps_goal` as fixed, are there any parameters that depend on it?

Reply:
Woah! I totally missed that. The equations I gave you uses tps to calculate the transaction capacity.  
```
let tps = 1

let block_window_duration = 300

let number_of_snark_workers = 32

let max_user_commands_per_block = tps * block_window_duration

let max_coinbases = 2

let max_fee_transfers = (Int.pow 2 (Int.ceil_log2 (max_user_commands_per_block + max_coinbases))) - 1

(*Scan state constants*)
let transaction_capacity_log_2 = 1 + Int.ceil_log2 (max_user_commands_per_block + max_coinbases) 

(* number of proofs required = number of transactions added - because they are bundled*)
let delay = (Int.pow 2 transaction_capacity_log_2) / number_of_snark_workers
```

Reply:
Probably some miscommunication. What I meant is we can calculate `transaction_capacity_log_2` using the tps instead of setting the  `transaction_capacity_log_2` after doing all the calculations for a given tps by hand

Reply:
Oh the calculations were removed in a later commit. Phew! It wasn't my imagination then. Anyway, we should still calculate `transaction_capacity_log_2` using tps and set the `delay` directly instead of calculating it using number of snark workers.

Reply:
In order to add *n* transactions we need ~*n* snark bundles. Transactions per second is necessarily proportional to snarks per second.

The number of snark bundles produced per second is proportional to the number of snark workers. The `work_delay` parameter doesn't affect how much work needs to be done in terms of throughput, only how much latency we can tolerate.

Reply:
I think you're right. If we're looking only for snark works for a specific block then delay will come into play but for a global number of snark workers required for every block then it wouldn't. And this number should be the global one. It should be `2^transaction_capacity_log_2 * (2*transaction_snark_proving_time/block_window_duration)`. Does that makes sense? This also means we need `block_window_duration >= 2*transaction_snark_proving_time` to achieve the tps goal